CONTRIBUTING.md

```diff
@@ -45,6 +45,9 @@ which uses [google-java-format](https://github.com/google/google-java-format) li
 
 ### Best practices that we follow
 
+* This project uses [semantic versioning](https://semver.org/). Except for major versions, a user should be able to update
+their dependency version on this project and have nothing break. This means we do not make breaking
+changes to the API (e.g., remove a public method) or to the ABI (e.g., change return type from void to non-void).
```

Comment:
Is the example for ABI really breaking? Because if that is breaking I think adding a new method is also breaking correct?

Also changing an argument from String -> CharSequence I thought it is fine, I know it breaks the sdk implementation but does it break an artifact that depends on small API version + new API + new SDK?

Reply:
So what you are saying is that we make a special exception for ABI compatibility to allow changes that would break SDK implementations but from which we hope that they do not break typical uses of the API package for instrumentation? We need to document that then.

Reply:
Adding new method is not breaking. Changing return type is breaking because `invokevirtual/invokeinterface` bytecoe instruction contain method descriptor to call. That descriptor contains return value. When method with such return value is not found when called, the `AbstractMethodError` is thrown.

Reply:
Note that *technically* implementing the API is just a special case of "using the API".

Reply:
I found a very good and detailed explication of the ABI compatibility https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html

Indeed changing the return type, or the type of an argument even if it is source compatible as in my example with the String is considered backwards incompatible.

Reply:
@Oberon00 have not read the entire document, but at first it seems that adding a new method to an interface (even without default) is considered backwards compatible, but that may break SDK implementation. I would not aim to fix that problem personally.

Reply:
Implementing the API is a special case, but that special case is not supposed to be included in other libraries so does not require ABI compatibility if I understand correctly

Reply:
Enforcing version coupling for API + SDK, including custom SDKs is ok I think though especially for the case of custom SDKs we need to declare that clearly (hopefully separate from this PR).

Still, ideally we can avoid it, and default interface methods help with that.